File: lib/ptabler/software/src/ptabler/expression/fuzzy.py
import typing
import polars as pl
import polars_ds as pds

from .base import Expression

StringDistanceMetric = typing.Literal[
    "levenshtein",
    "optimal_string_alignment",
    "jaro_winkler"
]

AnyExpression = Expression


class StringDistanceExpression(Expression, tag='string_distance'):
    """
    Computes string distance or similarity between two string expressions.
    Corresponds to StringDistanceExpression in TypeScript.
    Leverages functions from the polars-ds library.
    """
    metric: StringDistanceMetric
    """The specific distance metric to use (e.g., 'levenshtein')."""
    string1: 'AnyExpression'
    """The first string expression."""
    string2: 'AnyExpression'
    """The second string expression to compare against."""
    return_similarity: typing.Optional[bool] = False
    """
    If true, return similarity (0-1). If false (default), return distance.
    Jaro-Winkler always returns similarity.
    """

    def to_polars(self) -> pl.Expr:
        """Converts the expression to a Polars expression using polars-ds."""
        s1_polars = self.string1.to_polars()
        s2_polars = self.string2.to_polars()

        match self.metric:
            case "levenshtein":
                return pds.str_leven(
                    s1_polars, s2_polars, return_sim=self.return_similarity or False
                )
            case "optimal_string_alignment":
                return pds.str_osa(
                    s1_polars, s2_polars, return_sim=self.return_similarity or False
                )
            case "jaro_winkler":
                return pds.str_jw(
                    s1_polars, s2_polars
                )
            case _:
                # Should be unreachable due to Literal typing
                raise ValueError(
                    f"Unsupported string distance metric: {self.metric}")


FuzzyFilterDistanceMetric = typing.Literal['levenshtein', 'hamming']


class FuzzyStringFilterExpression(Expression, tag='fuzzy_string_filter'):
    """
    Filters strings based on their distance to a pattern string.
    Returns true if the distance is within the specified bound.
    Corresponds to FuzzyStringFilterExpression in TypeScript.
    Leverages functions from the polars-ds library.
    """
    metric: FuzzyFilterDistanceMetric
    """The distance metric to use ('levenshtein' or 'hamming')."""
    value: 'AnyExpression'
    """The string expression whose value will be compared."""
    pattern: 'AnyExpression'
    """The expression representing the string pattern to compare against."""
    bound: int
    """The maximum allowed distance for a match (inclusive). Must be non-negative."""

    def to_polars(self) -> pl.Expr:
        """Converts the expression to a Polars boolean expression using polars-ds."""
        if self.bound < 0:
            raise ValueError(
                f"FuzzyStringFilterExpression 'bound' ({self.bound}) cannot be negative.")

        value_polars = self.value.to_polars()
        pattern_polars = self.pattern.to_polars()

        match self.metric:
            case "levenshtein":
                return pds.filter_by_levenshtein(value_polars, pattern_polars, self.bound)
            case "hamming":
                return pds.filter_by_hamming(value_polars, pattern_polars, self.bound)
            case _:
                raise ValueError(
                    f"Unsupported fuzzy filter metric: {self.metric}")

File: lib/ptabler/schema/src/expressions/fuzzy.ts
import type { Expression } from './base';

/** Defines the supported string distance metrics. */
export type StringDistanceMetric =
  | 'levenshtein'
  | 'optimal_string_alignment'
  | 'jaro_winkler';

/**
 * Represents a string distance/similarity calculation between two expressions.
 * Computes metrics like Levenshtein, Optimal String Alignment, or Jaro-Winkler.
 */
export interface StringDistanceExpression {
  /** The type of operation, always 'string_distance'. */
  type: 'string_distance';
  /** The specific distance metric to use. */
  metric: StringDistanceMetric;
  /** The first string expression. */
  string1: Expression;
  /** The second string expression to compare against. */
  string2: Expression;
  /**
   * If true, the expression returns a similarity score (typically normalized between 0 and 1).
   * If false or undefined, it returns the raw edit distance (e.g., Levenshtein, OSA).
   * Jaro-Winkler inherently returns a similarity score; this flag might be ignored or influence its normalization if applicable.
   */
  returnSimilarity?: boolean;
}

/** Defines the supported fuzzy string filter distance metrics. */
export type FuzzyFilterDistanceMetric = 'levenshtein' | 'hamming';

/**
 * Represents a fuzzy string filter operation on an expression.
 * This operation compares the string value of an expression (`value`)
 * against another string or string expression (`pattern`) using a specified
 * distance metric (`levenshtein` or `hamming`), returning true if the distance is
 * within the specified `bound`.
 */
export interface FuzzyStringFilterExpression {
  /** The type of operation, always 'fuzzy_string_filter'. */
  type: 'fuzzy_string_filter';
  /** The distance metric to use for the fuzzy comparison. */
  metric: FuzzyFilterDistanceMetric;
  /** The expression whose string value will be compared. */
  value: Expression;
  /** The expression representing the string pattern to compare against. */
  pattern: Expression;
  /** The maximum allowed distance for a match (inclusive). */
  bound: number;
}
